lib/ipam/ipam.go

```diff
@@ -1383,6 +1343,91 @@ func (c ipamClient) convertBackendToIPAMConfig(cfg *model.IPAMConfig) *IPAMConfi
 	}
 }
 
+// ensureConsistentAffinity retrieves the pool and node for the given block and determines
+// if the pool still selects node. If it no longer matches, it will release the block
+// affinity for that node.
+// Returns a bool indicating if the block affinity was released.
+func (c ipamClient) ensureConsistentAffinity(ctx context.Context, b *model.AllocationBlock) (bool, error) {
+
+	// Retrieve node for this allocation. We do this so we can clean up affinity for blocks
+	// which should no longer be affine to this host.
+	host := getHostAffinity(b)
+	logCtx := log.WithFields(log.Fields{"cidr": b.CIDR, "host": host})
+
+	// If no hostname is found on the block affinity,
+	// there is no need to do an ip pool node selection check.
+	if host == "" {
+		logCtx.Debug("Block already has no affinity")
+		return false, nil
+	}
+
+	// If the IP pool which owns this block no longer selects this node,
+	// we should release the block's affinity to this node so it can be
+	// used elsewhere.
+	logCtx.Debugf("Looking up node labels for host affinity")
+	node, err := c.client.Get(ctx, model.ResourceKey{Kind: v3.KindNode, Name: host}, "")
+	if err != nil {
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
+			logCtx.WithError(err).WithField("node", host).Error("Failed to get node for host")
+			return false, err
+		}
+		logCtx.Info("Node doesn't exist, releasing IPs anyway")
```

Comment:
just spotted this, but we can't know for sure if this function will only be called in the "release IPs" context - we might want to do this elsewhere, so I might tweak this to say "Node doesn't exist, no need to release affinity".

Fine to do this in a follow-on PR. Actually, we shouldn't ever really hit this branch because if the node doesn't exist, we should have already released the affinity, I think. But, best to be safe!